# SQL Practice Problems

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [3]:
import pandas as pd
from sqlalchemy import create_engine

# Setting up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))#to expand the size of my container so it can take fill up my entire monitor

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [4]:
sql = """
SELECT COUNT(*)
FROM stores;
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [5]:
sql = """
SELECT * 
FROM stores
ORDER BY stores.name
LIMIT 4;
"""

pd.read_sql_query(sql, engine)

store                                name store_status  \
0   4986  218 Fuel Express & Chubby's Liquor            A   
1   3047          3047 Cub Foods / Iowa City            I   
2   3060               3060 Cub Foods / Ames            I   
3   3061         3061 Cub Foods / Sioux City            I   

                                       store_address address_info  
0  68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1  855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2  3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3  1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [6]:
sql = """
SELECT btl_price, bottle_qty
FROM sales
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor ASC
LIMIT 3;
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty
0     $8.99          12
1    $28.50           3
2    $28.50           1

<img src="./images/schema.png" alt="schema" width="750"/>

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [14]:
sql = """
SELECT SUM(sales.total)
FROM sales
WHERE sales."date" BETWEEN '01-31-2015' AND '03-01-2015';
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**:

$21295350.55

<img src="./images/schema.png" alt="schema" width="750"/>

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [61]:
sql = """
SELECT sales.vendor, COUNT(*) AS transactions, SUM(sales.bottle_qty) AS bottle_qty, SUM(sales.total) AS total
FROM sales
WHERE sales.category_name = 'IMPORTED VODKA'
GROUP BY sales.vendor
ORDER BY COUNT(*) DESC;
"""

df = pd.read_sql_query(sql, engine)
df.head()

vendor  transactions  bottle_qty       total
0  Pernod Ricard USA/Austin Nichols         42338      403450  7880365.88
1               Bacardi U.S.A. Inc.         26441      236047  5771242.07
2                   Diageo Americas         15725      152038  3346425.42
3   Constellation Wine Company Inc.         12565      116053  1895005.51
4                   Jim Beam Brands          5850      104119  1807097.23

**Prompt 5 Answer**: 

Pernod Ricard USA/Austin Nichols

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Create a column called `Markup` that shows the difference between the `btl_price` and the `state_btl_cost`. Take a screenshot and paste it into your notebook.

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [25]:
sql = """
SELECT btl_price - state_btl_cost as Markup
FROM sales
LIMIT 5;
"""

pd.read_sql_query(sql, engine)

markup
0  $5.75
1  $5.75
2  $5.75
3  $8.58
4  $2.50

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

> Note: To make this an easier question, you may take the simple average of percentage markups across all transactions for a store. You do **not** need to weight the average markup by the number of bottles purchased per transaction. If you want to practice your SQL skills, though, consider this a bonus challenge!

<img src="./images/schema.png" alt="schema" width="750"/>

In [7]:
sql = """
SELECT store, AVG((btl_price - state_btl_cost) / state_btl_cost) AS "Percentage Markup"
FROM sales
GROUP BY store
ORDER BY "Percentage Markup" DESC;
"""

pd.read_sql_query(sql, engine)

store  Percentage Markup
0      4013           0.525471
1      4024           0.521499
2      4266           0.518903
3      3833           0.515094
4      3660           0.514591
...     ...                ...
1347   9018           0.500000
1348   4784           0.499994
1349   4494           0.499909
1350   4534           0.499781
1351   4693           0.499214

[1352 rows x 2 columns]

**Prompt 7 Answer**: 

The store with the highest average markup percentage is store 4013, which has an average percentage markup of 52.5471%.

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [8]:
sql = """
SELECT sales.store, AVG((sales.btl_price - sales.state_btl_cost) / sales.state_btl_cost) AS "Percentage Markup", stores.name, stores.store_address
FROM sales 
    LEFT JOIN stores
    ON sales.store = stores.store
GROUP BY sales.store, stores.name, stores.store_address
ORDER BY "Percentage Markup" DESC
LIMIT 5;
"""

pd.read_sql_query(sql, engine)

store  Percentage Markup                        name  \
0   4013           0.525471      Wal-Mart 0841 / Tipton   
1   4024           0.521499  Wal-Mart 1546 / Iowa Falls   
2   4266           0.518903  Wal-Mart 1683 / Shenandoah   
3   3833           0.515094    Wal-Mart 3394 / Atlantic   
4   3660           0.514591   Wal-Mart 2935 / Knoxville   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...

**Prompt 8 Answer**:

- Wal-Mart 0841
- Wal-Mart 1546 
- Wal-Mart 1683
- Wal-Mart 3394
- Wal-Mart 2935

<img src="./images/schema.png" alt="schema" width="750"/>

### Prompt 9

Which ten counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the ten counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

In [9]:
sql = """
SELECT counties.county, (SUM(bottle_qty) / counties.population) AS "Bottles Sold per Capita", ROUND(AVG(products.bottle_size),1) AS "Average Bottle Size"
FROM counties
    LEFT JOIN sales
        LEFT JOIN products
        ON sales.item = products.item_no
    ON counties.county = sales.county
GROUP BY counties.county
ORDER BY "Bottles Sold per Capita" DESC
LIMIT 10;
"""

pd.read_sql_query(sql, engine)

county  Bottles Sold per Capita  Average Bottle Size
0      Dickinson                       22                974.1
1           Polk                       15                892.6
2     Black Hawk                       15                862.9
3    Cerro Gordo                       14                986.4
4        Johnson                       13                895.4
5          Scott                       13                871.6
6           Linn                       12                886.3
7  Pottawattamie                       11                893.7
8        Kossuth                       11               1041.0
9            Lee                       10                912.2

**Prompt 9 Answer**:

Kossuth County sells the largest average bottle size (among the ten counties selling the most liquor bottles per capita), with an average bottle size of 1041.0 milliliters.

<img src="./images/schema.png" alt="schema" width="750"/>

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [153]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(*)
FROM sales 
WHERE total::money != btl_price * bottle_qty
"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**:

There are no records where `total` is not equal to `btl_price * bottle_qty`.